mailpoet/lib/Automation/Integrations/MailPoet/MailPoetIntegration.php

```diff
@@ -39,5 +39,8 @@ public function register(Registry $registry): void {
     $registry->addSubject($this->subscriberSubject);
     $registry->addTrigger($this->segmentSubscribedTrigger);
     $registry->addAction($this->sendEmailAction);
+
+    // sync step args (subject, preheader, etc.) to email settings
+    $registry->onBeforeWorkflowStepSave([$this->sendEmailAction, 'saveEmailSettings']);
```

Comment:
Is there a reason we would want this as a beforeSave hook instead of an afterSave? I know we don't have an afterSave right now and I'm wondering if that's something we would want.

I'm thinking it's possible for things to go wrong between this hook and the workflow itself getting saved. If the workflow fails to save, do we even want the updated email to save?

Reply:
My thinking was a bit the opposite — instead of this:

> I'm thinking it's possible for things to go wrong between this hook and the workflow itself getting saved. If the workflow fails to save, do we even want the updated email to save?

I was thinking, "if the hook fails for some reason, do we want to have such a workflow saved"? It also seemed to me safer to first sync the changes to the newsletter entity and only then update the workflow so we don't have an "inconsistent" state where the workflow is saved, but some related data is not saved yet.

Reply:
I hadn't thought of it in the other direction like that. Excellent point!

I do still think the current implementation could potentially cause problems. 

Let's say I have a welcome email workflow. The trigger is signups to segment A, and the email subject is "Thanks for signing up for A!" I decide I don't want to send welcome emails for segment A anymore, but I do want to for segment B, so to be efficient I decide to edit the workflow instead of deleting it and creating a new one. I update the trigger to be for segment B, change the subject to be "Thanks for signing up for B!", and make some other change that somehow makes the workflow impossible to save. I see some error message saying, "Could not save workflow," and it happens to be the end of my work day so I sign off and tell myself I'll finish it tomorrow, not knowing that the email subject was actually changed. That evening 20,000 people sign up for segment A and get an email with the subject line, "Thanks for signing up for B!", and the very next day I get fired from my marketing job for incompetence.

Obviously this is an absurd and unlikely story. Is it a possibility in terms of the saving workflow right now though?

Part of me would love to wrap any database changes in a transaction so we could roll back if we encounter any errors, but that would certainly be overkill right now, and maybe the situation I outlined isn't even possible.

I think the current approach makes more sense than an `afterSave` so let's just keep this as is for now, though I am curious to hear if you think my scenario is possible or a concern at all.

Reply:
Yes, that absolutely makes sense to me. For now, doing it either way can cause these inconsistencies and it's rather a point of view which one is better or worse.

As far as the hooks go, I think we should provide both — before and after. But for our implementation of the send email step we can explore what would be the safest:
1. We can decide whether we consider before or after safer.
2. We can try to use a transaction. Do we use them in the plugin at all? 🤔 I know there were some issues such as some "perf. optimizing" plugins converting all tables to MyISAM and such. But maybe a query with a transaction doesn't fail on MyISAM, it just may not end up being transactionally performed. Not sure.
3. Maybe we can force-sync the args to the newsletter entity before sending. Then we should probably use "after" save, as you suggest, or maybe even no hook at all. Noting this to the unorganized checklist.

Reply:
Personally I like the idea of #3. Ensuring that we have the correct data in the newsletter at send time seems like the most important thing. And we definitely don't want to depend on transactions considering MyISAM, which I wasn't even thinking of. Another decision for the future!